contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+
+The external component watching the Kubernetes API on behalf of the external CSI volume driver must handle provisioning, deleting, attaching, and detaching.
+
+##### Provisioning and Deleting
+
+Provisioning and deletion operations are handled using the existing [external provisioner mechanism](https://github.com/kubernetes-incubator/external-storage/tree/master/docs), where the external component watching the Kubernetes API on behalf of the external CSI volume driver will act as an external provisioner.
+
+In short, to dynamically provision a new CSI volume, a cluster admin would create a `StorageClass` with the provisioner corresponding to the name of the external provisioner handling provisioning requests on behalf of the CSI volume driver.
+
+To provision a new CSI volume, an end user would create a `PersistentVolumeClaim` object referencing this `StorageClass`. The external provisioner will react to the creation of the PVC and issue the `CreateVolume` call against the CSI volume driver to provision the volume. The `CreateVolume` name will be auto-generated as it is for other dynamically provisioned volumes. The `CreateVolume` capacity will be take from the `PersistentVolumeClaim` object. The `CreateVolume` parameters will be passed through from the `StorageClass` parameters (opaque to Kubernetes). Once the operation completes successfully, the external provisioner creates a `PersistentVolume` object to represent the volume using the information returned in the `CreateVolume` call. The `PersistentVolume` object is bound to the `PersistentVolumeClaim` and available for use.
+
+To delete a CSI volume, an end user would delete the corresponding `PersistentVolumeClaim` object. The external provisioner will react to the deletion of the PVC and based on its reclamation policy it will issue the `DeleteVolume` call against the CSI volume driver commands to delete the volume. It will then delete the `PersistentVolume` object.
+
+##### Attaching and Detaching
+
+Attach/detach operations must also be handled by an external component (an “attacher”). The attacher watches the Kubernetes API on behalf of the external CSI volume driver for new VolumeAttachment objects (defined below), and triggers the appropriate calls against the CSI volume driver to attach the volume.
+
+More specifically, an external “attacher” must watch the Kubernetes API on behalf of the external CSI volume driver to handle attach/detach requests.
+
+Once the following conditions are true, the external-attacher should call `ControllerPublishVolume` against the CSI volume driver to attach the volume to the specified node:
+
+1. A new `VolumeAttachment` Kubernetes API objects is created by Kubernetes.
+2. The `VolumeAttachment.Spec.Attacher` value in that object corresponds to the name of the external attacher.
+3. The `VolumeAttachment.Status.IsAttached` value is not yet set to true.
```

Comment:
Do kubelets have any need to view VolumeAttachment objects, or update their status, or is interaction with that object limited to the attacher?

Reply:
Prefer `Attached` to `IsAttached` per API conventions

Reply:
> Do kubelets have any need to view VolumeAttachment objects, or update their status

Not now, maybe in future. We thought about moving `VolumesInUse` and `VolumesAttached` from `Node.Status` to individual AttachedVolume instances, we'll reconsider that in beta or later.

Reply:
>Prefer Attached to IsAttached per API conventions
￼ 
fixed